internal/provider/services/storage/client.go

```diff
@@ -169,3 +212,117 @@ func (d downloadShasumSignatureClient) SendMsg(m any) error {
 func (d downloadShasumSignatureClient) RecvMsg(m any) error {
 	return d.client.RecvMsg(m)
 }
+
+type uploadBinaryZipClient struct {
+	conn   *grpc.ClientConn
+	client services.Storage_UploadProviderBinaryZipClient
+}
+
+func (u *uploadBinaryZipClient) Send(request *provider.UploadProviderBinaryZipRequest) error {
+	return u.client.Send(request)
+}
+
+func (u *uploadBinaryZipClient) CloseAndRecv() (*provider.Response, error) {
+	defer func() { _ = u.conn.Close() }()
+	return u.client.CloseAndRecv()
+}
+
+func (u *uploadBinaryZipClient) Header() (metadata.MD, error) {
+	return u.client.Header()
+}
+
+func (u *uploadBinaryZipClient) Trailer() metadata.MD {
+	return u.client.Trailer()
+}
+
+func (u *uploadBinaryZipClient) CloseSend() error {
+	return u.client.CloseSend()
+}
+
+func (u *uploadBinaryZipClient) Context() context.Context {
+	return u.client.Context()
+}
+
+func (u *uploadBinaryZipClient) SendMsg(m any) error {
+	return u.client.SendMsg(m)
+}
+
+func (u *uploadBinaryZipClient) RecvMsg(m any) error {
+	return u.client.RecvMsg(m)
+}
```

Comment:
I've been looking at this hating the way that I original implemented it and has now been copied in lots of places.
Can you try the following, I think it should work
```suggestion
type uploadBinaryZipClient struct {
	conn   *grpc.ClientConn
	services.Storage_UploadProviderBinaryZipClient
}

func (u *uploadBinaryZipClient) Send(request *provider.UploadProviderBinaryZipRequest) error {
	return u.Storage_UploadProviderBinaryZipClient.Send(request)
}

func (u *uploadBinaryZipClient) CloseAndRecv() (*provider.Response, error) {
	defer func() { _ = u.conn.Close() }()
	return u.Storage_UploadProviderBinaryZipClient.CloseAndRecv()
}

```

You then instantiate the struct using:
```golang
     return &uploadBinaryZipClient{Storage_UploadProviderBinaryZipClient: upload, conn: conn}, nil
```